### DISCUSSION TITLE
Change the appearance of a comment

### DISCUSSION THREAD

**Discussion by ogurec-ogurec** [danirus/django-comments-xtd#244] [Q&A] [answered]
Hello Danirus!

I have a question: I would like to slightly change the look of the comment, for example like this:

![image](https://user-images.githubusercontent.com/66796268/102518712-f0ed0e80-4024-11eb-955e-2dcb494fc80c.png)

Name
date
Comment
like/dislike
Answer

Where can I change the html to change the comment template?

**Comment by danirus** [comment, marked as answer]
Hi @ogurec-ogurec , for the current version the answer depends on how you use the app. If you use the render_xtdcomment_tree tag the template to rewrite is `django_comments_xtd/comment_tree.html`. But if you use the ReactJS plugin, the code that renders the comment is in the `comment.jsx` file, in `static/django_comments_xtd/js/src/`.

And finally, when you don't use nested comments nor the ReactJS plugin, but the template tag `render_comment_list`, from the parent app [django-comments](https://github.com/django/django-contrib-comments). In this case the template used is `comments/list.html`. 

Based on the image you post I can see you use nested comments, so it's either `comment_tree.html` or the `comment.jsx` file.

**Reply by ogurec-ogurec** [reply]
@danirus Thanks for the answer, I tried to change the file static/django_comments_xtd/js/src/comment.jsx but nothing changed in rendering, what did I do wrong?

**Reply by danirus** [reply]
You have to run the webpack command to build the JavaScript plugin. Before that you need to clone the repository. I wrote [this page in the docs](https://django-comments-xtd.readthedocs.io/en/latest/javascript.html) about how to change the frontend. I recommend you to read it to setup the environment. Once you have the setup, run the `webpack --watch` command, it will have the effect of building the JavaScript plugin. The result is placed in the `django_comments_xtd/static/django_comments_xtd/js/` directory.

The process of changing the frontend is more tedious than simply rewriting Django templates. For v3 of the package I want to make that easier. But so far this is how it works. Please, don't hesitate to ask me if you get stuck on any step.

**Reply by ogurec-ogurec** [reply]
I did the following actions:

1. Installed git: https://github.com/danirus/django-comments-xtd.git

2. Changed the file:`django-comments-xtd-master\django_comments_xtd\static\django_comments_xtd\js\src\comment.jsx`

3. Run command from directory: django-comments-xtd-master
`npm install`

4. The directory was created: django-comments-xtd-master/**node_modules**

Which path to run the command from: webpack --watch?

**Reply by ogurec-ogurec** [reply]
Launched it in the directory, here is the log:

`I:\test django comments\new\django-comments-xtd-master\node_modules\.bin>webpack --watch`

```
webpack is watching the files:

Hash: e272681bed1e3e7da805a
Version: webpack 4.44.2
Time: 1034ms
Built at: 19.12.2020 21:31:05
 2 assets
Entrypoint plugin = plugin-2.8.1.js plugin-2.8.1.js.map
[0] external "React" 42 bytes {0} [built]
[1] external "ReactDOM" 42 bytes {0} [built]
[2] I:/test django comments/new/django-comments-xtd-master/django_comments_xtd/s
tatic/django_comments_xtd/js/src/commentbox.jsx 6.82 KiB {0} [built] [failed] [1
 error]
[3] I:/test django comments/new/django-comments-xtd-master/django_comments_xtd/s
tatic/django_comments_xtd/js/src/index.js 264 bytes {0} [built]

ERROR in I:/test django comments/new/django-comments-xtd-master/django_comments_
xtd/static/django_comments_xtd/js/src/commentbox.jsx
Module build failed (from ../babel-loader/lib/index.js):
SyntaxError: I:\test django comments\new\django-comments-xtd-master\django_comme
nts_xtd\static\django_comments_xtd\js\src\commentbox.jsx: Support for the experi
mental syntax 'jsx' isn't currently enabled (54:14):

  52 |                                  this.state.cids.length);
  53 |       var text = django.interpolate(fmts, [this.state.cids.length]);
> 54 |       return <h5 className="text-center">{text}</h5>;
     |              ^
  55 |     } else
  56 |       return "";
  57 |   }

Add @babel/preset-react (https://git.io/JfeDR) to the 'presets' section of your
Babel config to enable transformation.
If you want to leave it as-is, add @babel/plugin-syntax-jsx (https://git.io/vb4y
A) to the 'plugins' section to enable parsing.
    at Parser._raise (I:\test django comments\new\django-comments-xtd-master\nod
e_modules\@babel\parser\lib\index.js:748:17)
    at Parser.raiseWithData (I:\test django comments\new\django-comments-xtd-mas
ter\node_modules\@babel\parser\lib\index.js:741:17)
    at Parser.expectOnePlugin (I:\test django comments\new\django-comments-xtd-m
aster\node_modules\@babel\parser\lib\index.js:9116:18)
    at Parser.parseExprAtom (I:\test django comments\new\django-comments-xtd-mas
ter\node_modules\@babel\parser\lib\index.js:10569:22)
    at Parser.parseExprSubscripts (I:\test django comments\new\django-comments-x
td-master\node_modules\@babel\parser\lib\index.js:10150:23)
    at Parser.parseUpdate (I:\test django comments\new\django-comments-xtd-maste
r\node_modules\@babel\parser\lib\index.js:10130:21)
    at Parser.parseMaybeUnary (I:\test django comments\new\django-comments-xtd-m
aster\node_modules\@babel\parser\lib\index.js:10119:17)
    at Parser.parseExprOps (I:\test django comments\new\django-comments-xtd-mast
er\node_modules\@babel\parser\lib\index.js:9989:23)
    at Parser.parseMaybeConditional (I:\test django comments\new\django-comments
-xtd-master\node_modules\@babel\parser\lib\index.js:9963:23)
    at Parser.parseMaybeAssign (I:\test django comments\new\django-comments-xtd-
master\node_modules\@babel\parser\lib\index.js:9926:21)
    at Parser.parseExpressionBase (I:\test django comments\new\django-comments-x
td-master\node_modules\@babel\parser\lib\index.js:9871:23)
    at I:\test django comments\new\django-comments-xtd-master\node_modules\@babe
l\parser\lib\index.js:9865:39
    at Parser.allowInAnd (I:\test django comments\new\django-comments-xtd-master
\node_modules\@babel\parser\lib\index.js:11541:16)
    at Parser.parseExpression (I:\test django comments\new\django-comments-xtd-m
aster\node_modules\@babel\parser\lib\index.js:9865:17)
    at Parser.parseReturnStatement (I:\test django comments\new\django-comments-
xtd-master\node_modules\@babel\parser\lib\index.js:12045:28)
    at Parser.parseStatementContent (I:\test django comments\new\django-comments
-xtd-master\node_modules\@babel\parser\lib\index.js:11724:21)
    at Parser.parseStatement (I:\test django comments\new\django-comments-xtd-ma
ster\node_modules\@babel\parser\lib\index.js:11676:17)
    at Parser.parseBlockOrModuleBlockBody (I:\test django comments\new\django-co
mments-xtd-master\node_modules\@babel\parser\lib\index.js:12258:25)
    at Parser.parseBlockBody (I:\test django comments\new\django-comments-xtd-ma
ster\node_modules\@babel\parser\lib\index.js:12249:10)
    at Parser.parseBlock (I:\test django comments\new\django-comments-xtd-master
\node_modules\@babel\parser\lib\index.js:12233:10)
    at Parser.parseStatementContent (I:\test django comments\new\django-comments
-xtd-master\node_modules\@babel\parser\lib\index.js:11752:21)
    at Parser.parseStatement (I:\test django comments\new\django-comments-xtd-ma
ster\node_modules\@babel\parser\lib\index.js:11676:17)
    at Parser.parseIfStatement (I:\test django comments\new\django-comments-xtd-
master\node_modules\@babel\parser\lib\index.js:12030:28)
    at Parser.parseStatementContent (I:\test django comments\new\django-comments
-xtd-master\node_modules\@babel\parser\lib\index.js:11721:21)
    at Parser.parseStatement (I:\test django comments\new\django-comments-xtd-ma
ster\node_modules\@babel\parser\lib\index.js:11676:17)
    at Parser.parseBlockOrModuleBlockBody (I:\test django comments\new\django-co
mments-xtd-master\node_modules\@babel\parser\lib\index.js:12258:25)
    at Parser.parseBlockBody (I:\test django comments\new\django-comments-xtd-ma
ster\node_modules\@babel\parser\lib\index.js:12249:10)
    at Parser.parseBlock (I:\test django comments\new\django-comments-xtd-master
\node_modules\@babel\parser\lib\index.js:12233:10)
    at Parser.parseFunctionBody (I:\test django comments\new\django-comments-xtd
-master\node_modules\@babel\parser\lib\index.js:11221:24)
    at Parser.parseFunctionBodyAndFinish (I:\test django comments\new\django-com
ments-xtd-master\node_modules\@babel\parser\lib\index.js:11205:10)
    at Parser.parseMethod (I:\test django comments\new\django-comments-xtd-maste
r\node_modules\@babel\parser\lib\index.js:11155:10)
    at Parser.pushClassMethod (I:\test django comments\new\django-comments-xtd-m
aster\node_modules\@babel\parser\lib\index.js:12701:30)
    at Parser.parseClassMemberWithIsStatic (I:\test django comments\new\django-c
omments-xtd-master\node_modules\@babel\parser\lib\index.js:12590:12)
    at Parser.parseClassMember (I:\test django comments\new\django-comments-xtd-
master\node_modules\@babel\parser\lib\index.js:12532:10)
    at I:\test django comments\new\django-comments-xtd-master\node_modules\@babe
l\parser\lib\index.js:12477:14
    at Parser.withTopicForbiddingContext (I:\test django comments\new\django-com
ments-xtd-master\node_modules\@babel\parser\lib\index.js:11516:14)
 @ I:/test django comments/new/django-comments-xtd-master/django_comments_xtd/st
atic/django_comments_xtd/js/src/index.js 3:0-46 4:36-46
```

**Reply by ogurec-ogurec** [reply]
@danirus It seems that nothing worked for me, what can I do?)

**Reply by danirus** [reply]
That's not what I explained in the link I passed you above. Do you have any problem reading them?
Webpack is a JavaScript package that will get installed by running `npm install` in the root directory of the package. That is explained in the [JavaScript Plugin](https://django-comments-xtd.readthedocs.io/en/latest/javascript.html) page of the docs.
Please, follow those instructions.

**Reply by danirus** [reply]
Hi @ogurec-ogurec, something is wrong with the discussions, I got your last two messages after I sent my previous comment.
I see that you got it to work. But you seem to have an issue compiling the JavaScript code. I will check the error.

**Reply by danirus** [reply]
> @danirus It seems that nothing worked for me, what can I do?)

Did you run `npm install` to get all the packages listed in the `package.json` file?

**Reply by ogurec-ogurec** [reply]
yes, i run command: 

`(newenv) I:\test django comments\new\django-comments-xtd-master>npm install`


I get the following log:

```
npm WARN bootstrap@4.5.3 requires a peer of popper.js@^1.16.1 but none is instal
led. You must install peer dependencies yourself.
npm WARN django-comments-xtd-plugins@2.8.1 No repository field.

audited 767 packages in 5.736s

20 packages are looking for funding
  run `npm fund` for details

found 1 low severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
```

**Reply by danirus** [reply]
I can't say what's the issue. In both, Linux and Mac OS I don't get that error. You are using Windows, maybe there's some limitation there with babel and JSX. Did you try to fix it following the suggestion that is displayed below the issue?

> Add @babel/preset-react (https://git.io/JfeDR) to the 'presets' section of your
Babel config to enable transformation.

**Reply by ogurec-ogurec** [reply]
I don't quite understand where to add **@ babel / preset-response**

I run command: 
`npm install --save-dev @babel/preset-react`

being in the directory:  `(newenv) I:\test django comments\new\django-comments-xtd-master>`

but the result is the same(

**Reply by ogurec-ogurec** [reply]
As is usually the case, the error was in the most obvious place. After 4 days, I found it -  just had to use the command:`npx webpack`  and it worked.

In my case, the command: `webpack --watch` did not work.

Thank you for your help! @danirus

**Comment by ogurec-ogurec** [comment]
Can you please tell me how you can check that this is a child comment, and for example set the avatar class - "avatar-children". In other words, separate the rendering of the child comment and the normal comment? @danirus

**Reply by danirus** [reply]
In the `comment_tree.html` template, there is a `{% for item in comments %}`. Each `item.comment` is an instance of the `XtdComment`. So you can access the attributes of the `XtdComment` model directly from there. i.e.: you can do `{{ item.comment.level }}` to get the level. Level 0 is the root level of the tree. Nested comments always have a level > 0.

**Reply by ogurec-ogurec** [reply]
Thanks, as I understand it, `comment_tree.html` is if you use the `render_xtdcomment_tree` tag to render.

Is it possible to somehow add a class to nested comments in a ReactJS plugin?
